### PR TITLE
#161342860 Implement role-based access control functionality

### PR DESCRIPTION
**Pull request by victor-nweze-deel**
#### What does this PR do?
- *Implement role-based access control functionality*

#### Description of Task to be completed?
- Create a permissions config file
- Create a middleware to verify user rights
- Include middleware to a protected route
- Write integration test for the new middleware

#### How should this be manually tested?
- Git clone this repo
- Cd into the `haven-ah-backend` folder
- Start postman and fire a POST request to `localhost:500/api/v1/articles` , say, with both valid and invalid credentials

#### Screenshots (if appropriate)
screen shot 2018-11-12 at 8 12 42 am
screen shot 2018-11-12 at 8 13 20 am

#### What are the relevant pivotal tracker stories?
[#161342860](https://www.pivotaltracker.com/story/show/161342860)

### CODE REVIEW

File: server/middlewares/checkPermissions.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    return badHttpResponse(response, 403, 'ACCESS DENIED');
+  }
+
+  const { email, role } = user.dataValues;
+
+  if (role === 'admin' && email === process.env.SUPER_ADMIN_EMAIL) {
+    return next();
+  }
+
+  if (!(
```
Comment (TheoOkafor) [inline server/middlewares/checkPermissions.js:28]: Nice one. I like this part. 👍

**Review by TheoOkafor — COMMENTED** [review]
Nice work @ProfJigsaw

File: server/middlewares/checkPermissions.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  const resource = processUrl(request.url);
+
+  const user = await userRepo.getUserByParam('id', request.userId);
+  if (!user) {
+    return badHttpResponse(response, 403, 'ACCESS DENIED');
+  }
+
+  const { email, role } = user.dataValues;
+
+  if (role === 'admin' && email === process.env.SUPER_ADMIN_EMAIL) {
```
Comment (seyi-adeleke) [inline server/middlewares/checkPermissions.js]: why isnt this just `role === 'admin'`

File: server/utilities/processUrl.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import url from 'url';
+
+/**
```
Comment (seyi-adeleke) [inline server/utilities/processUrl.js:3]: Test this please

File: server/middlewares/checkPermissions.js (hunk 2 above)
Comment (victor-nweze-deel) [reply server/middlewares/checkPermissions.js]: We decided that a user could be upgraded to an admin but this doesnt give him superadmin privileges like deleting other admins or modifying the role of a superadmin. I am making a check for the superadmin (using the email is as good as using any other primary identifier, seeing as emails are unique). On instance of a superadmin, then we dont care what he was trying to do. He gets automatic access.

File: server/utilities/processUrl.js (hunk 3 above)
Comment (victor-nweze-deel) [reply server/utilities/processUrl.js:3]: I'll get right on this

Comment (victor-nweze-deel) [reply server/utilities/processUrl.js:3]: Done

**Comment by PriscillaSam** [comment]
This looks great. Well done @ProfJigsaw. 👍

File: server/migrations/20181114070235-update-user-model.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export default {
+  up: (queryInterface) => {
+    const query = 'DELETE FROM pg_enum '
```
Comment (seyi-adeleke) [inline server/migrations/20181114070235-update-user-model.js:3]: This is weird lol

**Review by seyi-adeleke — APPROVED** [review]

**Comment by seyi-adeleke** [comment]
Just checked the uncovered lines. Shouldn't be difficult to test.

File: server/migrations/20181114070235-update-user-model.js (hunk 4 above)
Comment (victor-nweze-deel) [reply server/migrations/20181114070235-update-user-model.js:3]: Yeah, somehow I have to manually delete the label from the meta table.

**Comment by victor-nweze-deel** [comment]
> Just checked the uncovered lines. Shouldn't be difficult to test.

I'm working on covering those lines

**Pull request merged**